vault/identity_store_oidc_provider.go

```diff
@@ -1060,3 +1179,461 @@ func (i *IdentityStore) pathOIDCProviderExistenceCheck(ctx context.Context, req
+	}
+
+	// Upsert the client in memdb
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+	if err := i.memDBUpsertClientInTxn(txn, client); err != nil {
+		i.logger.Debug("failed to upsert client in memdb", "error", err)
+		return client, nil
+	}
+	txn.Commit()
+
+	return client, nil
+}
+
+// memDBClientByID returns the client with the given ID from memdb.
+func (i *IdentityStore) memDBClientByID(id string) (*client, error) {
+	if id == "" {
+		return nil, fmt.Errorf("missing client ID")
+	}
+
+	txn := i.db.Txn(false)
+
+	return i.memDBClientByIDInTxn(txn, id)
+}
+
+// memDBClientByIDInTxn returns the client with the given ID from memdb using the given txn.
+func (i *IdentityStore) memDBClientByIDInTxn(txn *memdb.Txn, id string) (*client, error) {
+	if id == "" {
+		return nil, fmt.Errorf("missing client ID")
+	}
+
+	if txn == nil {
+		return nil, errors.New("txn is nil")
+	}
+
+	clientRaw, err := txn.First(oidcClientsTable, "id", id)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch client from memdb using ID: %w", err)
+	}
+	if clientRaw == nil {
+		return nil, nil
+	}
+
+	client, ok := clientRaw.(*client)
+	if !ok {
+		return nil, errors.New("unexpected client type")
+	}
+
+	return client, nil
+}
+
+// memDBClientByName returns the client with the given name from memdb.
+func (i *IdentityStore) memDBClientByName(ctx context.Context, name string) (*client, error) {
+	if name == "" {
+		return nil, fmt.Errorf("missing client name")
+	}
+
+	txn := i.db.Txn(false)
+
+	return i.memDBClientByNameInTxn(ctx, txn, name)
+}
+
+// memDBClientByNameInTxn returns the client with the given ID from memdb using the given txn.
+func (i *IdentityStore) memDBClientByNameInTxn(ctx context.Context, txn *memdb.Txn, name string) (*client, error) {
+	if name == "" {
+		return nil, errors.New("missing client name")
+	}
+
+	if txn == nil {
+		return nil, errors.New("txn is nil")
+	}
+
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	clientRaw, err := txn.First(oidcClientsTable, "name", ns.ID, name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch client from memdb using name: %w", err)
+	}
+	if clientRaw == nil {
+		return nil, nil
+	}
+
+	client, ok := clientRaw.(*client)
+	if !ok {
+		return nil, errors.New("unexpected client type")
+	}
+
+	return client, nil
+}
+
+// memDBDeleteClientByName deletes the client with the given name from memdb.
+func (i *IdentityStore) memDBDeleteClientByName(ctx context.Context, name string) error {
+	if name == "" {
+		return fmt.Errorf("missing client name")
+	}
+
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+
+	if err := i.memDBDeleteClientByNameInTxn(ctx, txn, name); err != nil {
+		return err
+	}
+
+	txn.Commit()
+
+	return nil
+}
+
+// memDBDeleteClientByNameInTxn deletes the client with name from memdb using the given txn.
+func (i *IdentityStore) memDBDeleteClientByNameInTxn(ctx context.Context, txn *memdb.Txn, name string) error {
+	if name == "" {
+		return nil
```

Comment:
should this also return Error `missing client name`?

Reply:
Done in 8f9893e.